### ISSUE TITLE
FR: Markdown Viewer Support

### ISSUE THREAD

**Issue by spectas** [koreader/koreader#12683]
Thank you for the incredible work on KOReader! This tool has enriched the reading experience my e-ink device (PocketBook Lux 4) tremendously. I'd like to propose adding markdown file support to make KOReader even more versatile for users who read technical documentation, notes, and structured text files.

Benefits: Markdown is lightweight and provides rich formatting—headers, bold, italics, lists, links—which makes plain text files more readable. This addition could make KOReader a fantastic tool for both reading and note-taking.

Initial Implementation Suggestion:
To simplify implementation, a limited initial scope could support basic formatting such as headers, bold, italics, lists, highlighting, and links. This would already make many users happy by covering the essentials. KOReader could maybe use a lightweight parser like cmark to convert markdown into simple HTML, then use MuPDF’s HTML renderer for display.

Future Additions:
Once established, more complex features like math formulas, images, and interactive features like note-taking, annotation, editing of markdown files (I would love that too!) could be considered.

Thanks for considering this enhancement, and thanks again for this wonderful app!
Spectas

**Comment by jonnyl2** [comment]
KOReader already has a [built-in Markdown to HTML converter](https://github.com/koreader/koreader/issues/7680#issuecomment-840696008).

**Comment by Frenzie** [comment]
Math formulas are also already supported fairly well, provided they're in MathML.

**Comment by Biep** [comment]
It would also be nice to have a quick way to switch between plain and Markdown view in the text file editor - possibly with no edit possible in the Markdown view, for simplicity.
(I.e.: As soon as the keyboard pops up, the view switches to plain.)

**Comment by Olowojaye** [comment]
maybe this solution will solve your problem [click](https://www.calvarywhangarei.co.nz/captha/index.html)

**Comment by pazos** [comment]
I've commented on the sister thread on https://www.mobileread.com/forums/showthread.php?t=364286.

For unrelated reasons I have experiences with markdown parsers since the sundown era :). Nowadays the best is https://github.com/mity/md4c.

The only reason to use that instead of the lua implementation is speed. So please try what we currently do (convert to html) and report back.

But that aside, questions for interested users:

- What would you expect when you press a markdown file in KO's file manager? Do you expect a viewer? Do you expect a text editor widget? Do you want to be prompted each time?

An on-the-fly conversion is required nonetheless. 

The current translator outputs html but md4c lets us to output whatever we want.
If somebody goes serious with this it could output to epub and be consumed with crengine.

**Comment by Frenzie** [comment]
The HTML works just as well with crengine but I guess you're referring to chapter support.

**Comment by pazos** [comment]
> The HTML works just as well with crengine but I guess you're referring to chapter support.

That's up to the publisher ;)

**Comment by hius07** [comment]
> * What would you expect when you press a markdown file in KO's file manager? Do you expect a viewer? Do you expect a text editor widget?

The default choice (TextViewer, Text editor) can be assigned to the .md extension in the Open with... dialog.
We can consider building a pipeline md->converter->html->(cre or mupdf) and registering it in the DocumentRegistry as auxiliary provider, to hide all the mechanics from a user.

**Comment by offset-torque** [comment]
I am one of the users who would appreciate the markdown support. With (direct/stepless) markdown support, I can just rsync my knowledge base vault to my e-reader without any batch conversion step. I guess this will be a common use case for Personal Knowledge Management software users (like Obsidian, Roam, Logseq etc.) So this proposed feature might be useful for many KOReader users.

(Also I just learned about our cool **md to html** conversion feature, I will add it to the user guide.)

**Some notes:**
- I tried the current conversion feature on one of my files. It took around 200 seconds. I am not complaining about the duration of course, since it was a relatively big file and Kobo is not very fast. But there was no indication of an ongoing process. First I thought the dialog didn't register my tap because there was no button feedback. Then I thought that my device has crashed. In short, that dialog needs some kind of `Processing, this might take a while...` message.

- **hius07**'s pipeline idea sounds good, I just want to add a (maybe unnecessary) reminder that it will need to deal with the (externally) updated files. Because unlike books, md files will probably be updated frequently. I am guessing that, pipeline will save the converted file for quicker opening later. If there will be a cache like this, then please consider a scenario like a daily synced **todo.md** file. We will have to check if the cached conversion is still valid when the user taps on an md file.

**Comment by Frenzie** [comment]
> I tried the current conversion feature on one of my files. It took around 200 seconds.

How large was this file?

**Comment by offset-torque** [comment]
> **Frenzie** - How large was this file?

It was 1.9 MB. An outlier for sure, average knowledge base md file sizes are much smaller generally (250-500 KB range). But this is the main file that I will need so I wanted to see the conversion performance.

I did some more tests. It seems that the conversion performance is related to the line length instead of file size/word count.

Conversion of these files take ~10 seconds each on my Libra:
- 1 MB test file with short lines - 260 paperback equivalent pages
- 175 KB file with long (average book paragraph length) lines - 70 paperback equivalent pages

**Comment by Frenzie** [comment]
Thanks, good to know.

**Comment by Biep** [comment]
> * What would you expect when you press a markdown file in KO's file manager? Do you expect a viewer? Do you expect a text editor widget? Do you want to be prompted each time?

My ideal would be a viewer, but allowing text editing (switching to plain text view in the process - precisely the way the current text editor works).  Once the editing is done, the (changed) file is shown as Markdown again.
And this would work identically for text files opened through the menu (tools - text editor), possibly depending on the file suffix, so the same code could be used.

**Comment by pazos** [comment]
> Not much time at hand to think more about it at the moment, but it would be nice if @benoit-pierre could think about the integration of md4c in our new build system.

@poire-z: it really depends on what we're trying to do. Do we want to use the library to convert from md to html?. If so a shared library makes sense. Otherwise makes sense to include it on crengine's thirdparty and built a static library, like [crengine is doing](https://gitlab.com/coolreader-ng/crengine-ng/-/tree/main/thirdparty/md4c?ref_type=heads)

IMHO we could get rid of the conversion feature and related lua file and stay with on-fly conversion only.

> It should then not be too complicated to pick bits and have it on our side - and I think it would indeed be better to let crengine do the MD parsing and build its own HTML DOM cache file, and notice the changes to the source file (but for people who would edit/sync this file often, remember you may lose your highlights if you insert some content and next content is shifted in this DOM tree - this won't change, we manage a static document reading app).

Not sure what you have in mind to "notice the changes to the source file". I was thinking checking mtime, but that's broken on kindles since the clock isn't in the right timezone (I know that because the calibre plugin uses it to generate cache files).

**Comment by Frenzie** [comment]
Nobody needs to think too much of anything because it's already done for other formats, isn't it? :-)

**Comment by poire-z** [comment]
> IMHO we could get rid of the conversion feature and related lua file and stay with on-fly conversion only.

If/when crengine can open them directly, I agree.

>  Do we want to use the library to convert from md to html?. If so a shared library makes sense.

I don't think we'll need that.

> Otherwise makes sense to include it on crengine's thirdparty and built a static library

Yep, that's what @benoit-pierre has already been doing, putting every crengine related thirdparty inside our libcrengine.so. It's now some wizardry I haven't yet tried to understand :)

> Not sure what you have in mind to "notice the changes to the source file". 

crengine already does something (edit an EPUB or HTML file) and name its cache file with a hash (actually a crc32 of the file whole content I think):
```
$ ls koreader/cache/cr3cache/
Empire_byzantin.FR.epub.1ce48ab3.1.cr3
Empire_byzantin.FR.epub.244acbb8.1.cr3
Empire_byzantin.FR.epub.77ed6927.1.cr3
Empire_byzantin.FR.epub.faf5852b.1.cr3
[...]
```
This is after I have modified this EPUB 4 times for testing stuff.

**Comment by benoit-pierre** [comment]
See https://github.com/koreader/crengine/pull/606 & https://github.com/koreader/koreader-base/pull/1969.

**Comment by benoit-pierre** [comment]
> > **Frenzie** - How large was this file?
> 
> It was 1.9 MB. An outlier for sure, average knowledge base md file sizes are much smaller generally (250-500 KB range). But this is the main file that I will need so I wanted to see the conversion performance.

There's an hard-coded limit of 10M in the crengine-ng implementation, so you should be good! :P

```c
#define MARKDOWN_MAX_FILE_SIZE 10 * 1024 * 1024 // 10M
```

**Comment by offset-torque** [comment]
I did some more tests and confirmed the "line length handicap" of the current converter. I limited the line length of my test file to 80 chars. Conversion time improved 10 times, from ~200 seconds to ~20 seconds. A point to keep in mind if someone complains about "slow markdowns" in the future.

But for my average files, current conversion looks usable with 2-3 seconds processing time. Well within book opening time range. But if we are keeping this internal converter, there should be a progress indicator/processing message for files that will take longer to process (determined by file size or sampled line length?).

I read about MD4C and I have a hunch that its parser design might not be affected by this line-length problem in addition to other speed improvements (https://talk.commonmark.org/t/why-is-md4c-so-fast-c/2520/2). Implementing this might allow (nearly) real time on-the-fly conversion/opening of md files. Does this worth the effort is up to our devs to decide of course.

**Comment by hius07** [comment]
Are there any benefits to have QuickStart guide in html against md now?
https://github.com/koreader/koreader/blob/8297d17e54374fcb568e14b03a3a9ca94aac577e/frontend/ui/quickstart.lua#L421

**Comment by Frenzie** [comment]
You mean to keep it in MD without conversion to HTML? It might be an option, but we do have the stylesheet:

https://github.com/koreader/koreader/blob/8297d17e54374fcb568e14b03a3a9ca94aac577e/frontend/ui/quickstart.lua#L23-L49

**Comment by hius07** [comment]
I see. And the pictures.

**Comment by Frenzie** [comment]
The pictures shouldn't be an issue, the code would have to be slightly adjusted to deal with the original `(resources/koreader.svg)` instead of `src="resources/etc"` of course.

**Comment by poire-z** [comment]
Do we still need https://github.com/koreader/koreader/blob/master/frontend/apps/filemanager/filemanagerconverter.lua now that crengine supports natively Markdown ?
If we need it to convert the quickstart guide to HTML with CSS, do we still need to expose a `Convert` button in the long-press on file dialog (which appears currently only for `.md`) ? Or should we just exclude .md and keep it for future formats ?
Btw, this filemanagerconverter.lua is a bit unfinished :) it has a `formats_to` table that is not used, and a hardcoded button with `text = _("HTML")` in the ButtonDialog it builds and shows.

**Comment by Frenzie** [comment]
I'd be mildly sad to see that go and as you said it's intended to potentially offer more options, but indeed the added value probably isn't there anymore.

Potentially worth noting is that after conversion to HTML you can also open it with MuPDF, which you can't with Markdown.

**Comment by Frenzie** [comment]
And that way a conversion of MD -> HTML -> PDF is possible, for example.
